### PR TITLE
[#156] Continuous Deployment with Heroku

### LINKED ISSUE #146: App should run when component pieces are on different systems
Right now, a lot of our code is hardcoded to specific URLs or to localhost. For example, the URL of the server is hardcoded to Localhost in the sample apps and in the frontend. We should fix this so the url is dynamic / configurable.
- For frontend: see what the standard is for specifying the URL of the backend in a react app
- For sample apps: use a configuration file where we can set the URL of the server

### LINKED ISSUE #156: Set up continuous deployment
We should deploy the Master branch version of our code to a "production" environment in which we can collect usage data.
We will have to send an email to concordia IT, CCing the professor, so they can provide us with a computer or two to deploy onto. We will then have to set up a github action to deploy to that machine.

We should also register a domain name & get a certificate for it so we can test the app with HTTPS.

**Acceptance criteria**:

- Switchboard application is deployed in a production environment
- All functionality should work as in development environment
- Deploy a new version of the application every time a branch is merged to master

### PR DESCRIPTION
**Pull request by ButlerJAL**
# General info
Continuous deployment using Heroku will help us improve our productivity. 

**Issue number**: #156 

**Task description**: 

 - Configure Heroku and Github for continuous automatic deployments of master
 - Configure a Heroku pipeline to help with reviewing PRs (Review apps)
 - Invite all collaborators to Heroku
 - Reduce the tediousness of manually configuring back-and-forth between development and production environments

# Testing

**Test coverage**: n/a

# Additional notes

**Note to reviewers**:
Must be merged to master to trigger gh-pages deploy action

**To do before merging**: n/a

~~**To do after merging**: Update frontend application http requests to use deployed production app~~

### CODE REVIEW

**Comment by ButlerJAL** [comment]
CI failure because SonarCloud code analysis detecting a lack of code coverage on the DataSourceConfig class I added in this PR to programmatically set our DataSource configurations using Spring Profiles. Overall coverage after merge stays at above 90%. I'm inclined to ignore this failure as I feel that coverage on a configuration class is a little much and wouldn't be too sure on how to go about it. Team, let me know what you think?

File: src/main/java/org/beanpod/switchboard/config/DataSourceConfig.java
Comment (felixlapierre) [inline src/main/java/org/beanpod/switchboard/config/DataSourceConfig.java:22]: What are these credentials from? Are they even used?

File: src/main/java/org/beanpod/switchboard/config/DataSourceConfig.java
Comment (felixlapierre) [inline src/main/java/org/beanpod/switchboard/config/DataSourceConfig.java]: Do you know if these values can still be overridden with command line arguments?  (I can check but i don't have time right now)

**Review by felixlapierre — DISMISSED** [review]
Putting my approval, but i'd like us to check that we can still override username/password by passing command-line arguments to spring-boot:run, since that's the solution i've been using that lowers the odds of accidentally committing passwords.

I agree with @MrJCipherButtles that we can merge this even though coverage is flagged as insufficient, this isn't really code that can be unit tested.

File: src/main/java/org/beanpod/switchboard/config/DataSourceConfig.java
Comment (ButlerJAL) [reply src/main/java/org/beanpod/switchboard/config/DataSourceConfig.java:22]: These are the default h2 database credentials. It's common practice to use the defaults for a testing db.

File: src/main/java/org/beanpod/switchboard/config/DataSourceConfig.java
Comment (ButlerJAL) [reply src/main/java/org/beanpod/switchboard/config/DataSourceConfig.java]: Yes we may, but I have not seen usage of args in our code. Did you program the use of cli arguments locally on your end? If so, what arg names did you use or would want to use?

Comment (ButlerJAL) [reply src/main/java/org/beanpod/switchboard/config/DataSourceConfig.java]: Oh, my guess is that you were prefixing your cli args with '-' which adds the property to the Spring Environment effectively overriding application.properties. I'll push an adaptation in a minute to mock the same behaviour programmatically to not impede your flow

**Comment by mohhef** [comment]
> CI failure because SonarCloud code analysis detecting a lack of code coverage on the DataSourceConfig class I added in this PR to programmatically set our DataSource configurations using Spring Profiles. Overall coverage after merge stays at above 90%. I'm inclined to ignore this failure as I feel that coverage on a configuration class is a little much and wouldn't be too sure on how to go about it. Team, let me know what you think?

@MrJCipherButtles I would execlude the whole config package from sonar testing im the pom

File: src/main/java/org/beanpod/switchboard/config/DataSourceConfig.java
Comment (ButlerJAL) [reply src/main/java/org/beanpod/switchboard/config/DataSourceConfig.java]: Done.

**Comment by ButlerJAL** [comment]
> > CI failure because SonarCloud code analysis detecting a lack of code coverage on the DataSourceConfig class I added in this PR to programmatically set our DataSource configurations using Spring Profiles. Overall coverage after merge stays at above 90%. I'm inclined to ignore this failure as I feel that coverage on a configuration class is a little much and wouldn't be too sure on how to go about it. Team, let me know what you think?
> 
> @MrJCipherButtles I would execlude the whole config package from sonar testing im the pom

I guess so, done!

**Review by mohhef — DISMISSED** [review]
LGTM

**Comment by ButlerJAL** [comment]
@mohhef since your last approval, I added an app.json file that helps with the configuration of automatically created Heroku review apps. In summary, every PR will attempt to deploy a Heroku review app and connect it to its own remote MySQL database instance, so, the need of setting up a local development environment to run the backend is no longer needed. Have a look when possible, and kindly re-approve if all seems good, Thanks!

**Review by mohhef — DISMISSED** [review]
reapprove, from my understanding this does not change any current development flow on our local machines correct?

**Comment by ButlerJAL** [comment]
> reapprove, from my understanding this does not change any current development flow on our local machines correct?

@mohhef, if I can help it, I won't stale anymore of your approvals as this should safely be the last re-review request lol, thank you.

**BACKEND:**
Development flow using cli arguments as brought up by @felixlapierre is still possible using the `development` active Spring Profile. If you're not a fan of using a cli, there's an easy gui way to do so as well through the intellij IDE run configurations window:
Screen Shot 2021-01-08 at 12 03 47 PM
Note, you do not need to edit application.properties nor DataSourceConfig.java directly which effetively reduces the likelihood of pushing user-sensitive configurations to the repo. Simply type "development" in the active profiles section of the run configuration, and specify zero or more of the following program arguments:
- --spring.datasource.username (default: root, when omitted)
- --spring.datasource.password (default: root, when omitted)
- --spring.datasource.url (default: jdbc:mysql://localhost:3306/switchboard?allowPublicKeyRetrieval=true&useSSL=false&serverTimezone=UTC, when omitted)

**FRONTEND:**
Also, in similar fashion, I extracted URLs from the code and created custom environment variables to help seamlessly set the proper environment-specific settings with minimal to no configuration when running our React app (frontend).
Mind you with Heroku, you can spin up a non-production backend review app for your PRs, so, you may change only the DOMAIN environment variable within `.env.development` to have your development react app (frontend) point to a backend review app instead of a local backend app.


Overall, this should greatly improve contributor experience and productivity as you can now pull and push the project without messing around with setting and reverting configurations for development and production. 

P.S. On a production environment note, our GitHub Pages will run our production-ready frontend, and will always point to our production-ready backend hosted on an always online Heroku Hobby Dyno, so, presenting our app can be as simple as providing the link to our GitHub Pages. Both production frontend and backend derive from our master branch.

**Review by mohhef — DISMISSED** [review]
Reapproving stale.  Great, so in my case where i just load the project and run it from intellij it should use the defaults.

File: app.json
Comment (felixlapierre) [inline app.json]: Change the name to `switchboard`

File: frontend/.env.test
Comment (felixlapierre) [inline frontend/.env.test:4]: Does this grab from system environment variables if they are defined? If so, we should prefix these with `SWITCHBOARD_APP` to avoid conflicts. If not, forget about it.
Ex. `SWITCHBOARD_APP_DOMAIN`, `SWITCHBOARD_APP_ENCODER`, etc.

**Review by felixlapierre — DISMISSED** [review]
Minor suggestions, either merge & do the fixes later or just fix these and i'll re-approve

File: frontend/.env.test
Comment (ButlerJAL) [reply frontend/.env.test:4]: Custom environment variables must start with `REACT_APP_`, so, it rules out the possibility of the proposed suggestion.

File: app.json
Comment (ButlerJAL) [reply app.json]: I’ll make the change within the next 2 hours 👍

Comment (ButlerJAL) [reply app.json]: Done! Was able to do it on my phone 📱

**Review by felixlapierre — APPROVED** [review]

**Review by mohhef — APPROVED** [review]

**Comment by felixlapierre** [comment]
> Development flow using cli arguments as brought up by @felixlapierre is still possible using the `development` active Spring Profile. If you're not a fan of using a cli, there's an easy gui way to do so as well through the intellij IDE run configurations window:

Just writing it down so i don't forget: if you want to override the application properties it doesn't seem to work by adding VM options, you have to enter them in the "Override parameters" section under "active profiles"

**Pull request merged**
